nbdime/diffing/notebooks.py

```diff
@@ -133,23 +108,69 @@ def compare_output_data(x, y):
     return True
 
 
-# Keeping these here for the comments and as as a reminder for possible future extension points:
-def __unused_diff_single_outputs(a, b, path="/cells/*/output/*"):
+def diff_single_outputs(a, b, path="/cells/*/output/*",
+                        predicates=None, differs=None):
     "DiffOp a pair of output cells."
     assert path == "/cells/*/outputs/*"
+    assert a.output_type == b.output_type
+
     # TODO: Handle output diffing with plugins? I.e. image diff, svg diff, json diff, etc.
-    # FIXME: Use linebased diff of some types of outputs:
-    # if a.output_type in ("execute_result", "display_data"):
-    #    a.data.key if key.startswith('text/') or key in _non_text_split_mimes = {
-    #        'application/javascript','image/svg+xml'}
-    #    a.text
-    return diff(a, b)
-def __unused_diff_source(a, b, path, predicates, differs):
+    if a.output_type in ("execute_result", "display_data"):
+        di = MappingDiffBuilder()
+
+        a_conj = copy.deepcopy(a)
+        del a_conj['data']
+        b_conj = copy.deepcopy(b)
+        del b_conj['data']
+        dd_conj = diff(a_conj, b_conj)
+        if dd_conj:
+            for e in dd_conj:
+                di.append(e)
+
+        dd = diff_mime_bundle(a.data, b.data, path=path+"/data")
+        if dd:
+            di.patch("data", dd)
+
+        return di.validated()
+    else:
+        return diff(a, b)
+
+
+def diff_mime_bundle(a, b, path=None):
+    di = MappingDiffBuilder()
+
+    akeys = set(a.keys())
+    bkeys = set(b.keys())
+    # Sorting keys in loops to get a deterministic diff result
+    for key in sorted(akeys - bkeys):
+        di.remove(key)
+
+    # Handle values for keys in both a and b
+    for key in sorted(akeys & bkeys):
+        avalue = a[key]
+        bvalue = b[key]
+
+        # FIXME: Use linebased diff of some types of outputs:
+        # if key.startswith('text/') or key in _non_text_split_mimes = {
```

Comment:
Maybe put a collection of mime-type patterns to test as a module global, so it's easier to add to. What you have here is a reasonable starting point.